src/Shared/test/Shared.Tests/UrlDecoderTests.cs

```diff
@@ -0,0 +1,199 @@
+        }
+
+        [Theory]
+        [InlineData("%2F")]
+        public void ByteFormsEncodingDecodeInPlaceDecodesPercent2F(string input)
+        {
+            var source = Encoding.UTF8.GetBytes(input.ToCharArray());
+            var length = UrlDecoder.DecodeInPlace(source.AsSpan(), true);
+            Assert.Equal(1, length);
+            Assert.True(source.AsSpan(0, length).SequenceEqual(Encoding.UTF8.GetBytes("/").AsSpan()));
+        }
+
+        [Theory]
+        [InlineData("%FF%FF%FF%FF")] // FF invalid first byte
+        [InlineData("%F7%BF%BF%BF")] // beyond 0x10FFFF
+        [InlineData("%F7%C0")] // Following byte does not start with 10xx xxxx
+        [InlineData("%F0%81")] // Not enough bytes
+        [InlineData("%ED%A0%81")] // Invalid range 0xD800-0xDFFF
+        public void StringOutOfUtf8RangeDecodeInPlaceLeavesUnencoded(string input)
+        {
+            var source = input.ToCharArray();
+            var length = UrlDecoder.DecodeInPlace(source.AsSpan());
+            Assert.Equal(input.Length, length);
+            Assert.True(source.AsSpan(0, length).SequenceEqual(input.AsSpan()));
+        }
+
+        [Theory]
+        [InlineData("%FF%FF%FF%FF")] // FF invalid first byte
+        [InlineData("%F7%BF%BF%BF")] // beyond 0x10FFFF
+        [InlineData("%F7%C0")] // Following byte does not start with 10xx xxxx
+        [InlineData("%F0%81")] // Not enough bytes
+        [InlineData("%ED%A0%81")] // Invalid range 0xD800-0xDFFF
+        public void ByteOutOfUtf8RangeDecodeInPlaceLeavesUnencoded(string input)
+        {
+            var source = Encoding.UTF8.GetBytes(input.ToCharArray());
+            var length = UrlDecoder.DecodeInPlace(source.AsSpan(), true);
+            Assert.Equal(source.Length, length);
+            Assert.True(source.AsSpan(0, length).SequenceEqual(Encoding.UTF8.GetBytes(input).AsSpan()));
+        }
+
+        public static IEnumerable<object[]> PathTestData
+        {
+            get
+            {
+                return new List<object[]>()
+                {
+                    new[] { "hello", "hello" },
+                    new[] { "/", "/" },
+                    new[] { "http://localhost:5000/api", "http://localhost:5000/api" },
+                    new[] { "/api/abc", "/api/abc" },
+                    new[] { "/api/a%2Fb", "/api/a%2Fb" },
+                    new[] { "/a%20b", "/a b" },
+                    new[] { "/a%24b", "/a$b" },
+                    new[] { "/a%C2%A2b", "/a¢b" },
+                    new[] { "/a%E0%A4%B9b", "/aहb" },
+                    new[] { "/a%E2%82%ACb", "/a€b" },
+                    new[] { "/a%ED%95%9Cb", "/a한b" },
+                    new[] { "/a%F0%90%8D%88b", "/a𐍈b" },
+                    new[] { "/a%25b", "/a%b" },
+                    new[] { "/%E4%BD%A0%E5%A5%BD", "/你好" },
+                    new[] { "/a%%2Fb", "/a%%2Fb" },
+                    new[] { "/a%2Fb+c", "/a%2Fb+c" },
+                };
```

Comment:
Consider also adding a test:

```
/%C3%C3%A1 -> /%C3á
```

This would help detect regressions where `UrlDecoder` tries to consume too many bytes. In the case above, we don't want the first `C3` to consume the second `C3`, because the second `C3` is part of the pair `[C3 A1]`, which is a valid standalone char `'á'`.

Reply:
I added the test.